solr/bin/solr

```diff
@@ -2158,6 +2158,35 @@ function mk_writable_dir() {
   fi
 }
 
+# Check whether at least one of lsof/netstat/ss are available
+function get_port_tool() {
+  for tool in 'lsof' 'netstat' 'ss'; do
+      which $tool &> nul
+      if [[ $? -eq 0 ]]; then
+        echo $tool
+        return
+      fi
+  done
+}
+
+# Will get the PID of the java process running on the specified port, if one is running
+# Otherwise empty result. Use get_port_tool to check whether lsof/ss/netstat is installed first
+function check_port_in_use() {
+  local port=$1
+
+  case $(get_port_tool) in
+    lsof)
+      lsof -iTCP:${port} 2>/dev/null | grep "LISTEN" | grep -Eow "java\s*?[0-9]+" | grep -Eow [0-9]+
+      ;;
+    ss)
+      ss -ntpa 2>/dev/null | grep ":${port} " | grep 'LISTEN ' | grep -Eow "\"java\",(pid=)?[0-9]+?" | grep -Eow "[0-9]+"
+      ;;
+    netstat)
+      netstat -nlp 2>/dev/null | grep ":${port} " | grep -Eow " [0-9]+/java" | grep -Eow "[0-9]+"
+      ;;
```

Comment:
Analogous to the change above (restoring the more targeted use of `lsof`, avoiding the need for `grep`), I wonder if there are more targeted (but still portable) versions of this command for `ss` and `netstat`?

Reply:
My thinking here was that it's better to verify that the process that's listening is a java process, rather than some other process that is running the port you expect Solr on.

However, that's partly because my original use of these scripts was in a more general context. In this context, I guess we already have confidence that the port wasn't already in use?

Reply:
Right yep. 

I see `ss` supports filtering to listening sockets and filtering by ports - `ss -l -at '( sport = :8983 )'`, but I don't know whether those options have been standard for a long time or not, but I also can't see anything suggesting that they're new.
The `netstat` `-l` is already filtering to listen, but it doesn't seem like there's an explicit port filter for it.

I'll change the `ss` command to use those options

Reply:
Great! thanks for looking into it. This seems reasonable, and maybe if anyone else wants to weigh in wrt how standard these flags are, that could be helpful.

Assuming continuing to use grep, we really don't _need_ "extended" grep here (`grep -E` -- nor [here](https://github.com/apache/solr/blob/e35347e7cc12d91af91cb38f9cef43ccc7a7b40e/solr/bin/solr#L180)). Removing the `-E` flag would just mean a little bit of extra escaping of special characters in the pattern -- not sure it's worth it, totally fine as-is, just thought I'd mention.

Reply:
Thought I would try it anyway. 
Also hadn't realized that I had lost the `-sTCP:LISTEN` filter from lsof, so that's back in :)

Reply:
Ah that's a pain. Yeah, on linux it gives the PID/program name column
```
# netstat -nlp 2>/dev/null
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:8983          0.0.0.0:*               LISTEN      22862/java
tcp        0      0 0.0.0.0:9983            0.0.0.0:*               LISTEN      3380/java
tcp        0      0 127.0.0.1:7983          0.0.0.0:*               LISTEN      22862/java
```

https://apple.stackexchange.com/a/327967 suggests `-v` is the nearest alternative.